File: src/services/recording.ts
import { invoke } from "@tauri-apps/api/core";
import type { RecordingConfig, RecordingStatus, RecordingStats } from "@/types";

export async function startRecording(
  region?: { x: number; y: number; width: number; height: number }
): Promise<void> {
  return invoke<void>("start_recording", { region });
}

export async function stopRecording(): Promise<string> {
  return invoke<string>("stop_recording");
}

export async function pauseRecording(): Promise<void> {
  return invoke<void>("pause_recording");
}

export async function resumeRecording(): Promise<void> {
  return invoke<void>("resume_recording");
}

export async function getRecordingStatus(): Promise<RecordingStatus> {
  return invoke<RecordingStatus>("get_recording_status");
}

export async function getRecordingStats(): Promise<RecordingStats> {
  return invoke<RecordingStats>("get_recording_stats");
}

export async function getRecordingConfig(): Promise<RecordingConfig> {
  return invoke<RecordingConfig>("get_recording_config");
}

export async function updateRecordingConfig(config: Partial<RecordingConfig>): Promise<void> {
  return invoke<void>("update_recording_config", { config });
}

export async function addRecordingFrame(frameData: string): Promise<void> {
  return invoke<void>("add_recording_frame", { frameData });
}

export async function checkFfmpeg(): Promise<{ available: boolean; version?: string; path?: string }> {
  return invoke<{ available: boolean; version?: string; path?: string }>("check_ffmpeg");
}

File: src-tauri/src/commands/recording.rs
//! Recording commands

use crate::models::recording::{RecordingConfig, RecordingStats, RecordingStatus};
use crate::services::recording::RecordingService;
use std::sync::Mutex;
use tauri::command;

pub struct RecordingState {
    pub service: Mutex<RecordingService>,
}

impl RecordingState {
    pub fn new() -> Self {
        Self {
            service: Mutex::new(RecordingService::new()),
        }
    }
}

impl Default for RecordingState {
    fn default() -> Self {
        Self::new()
    }
}

#[command]
pub fn start_recording(
    config: Option<RecordingConfig>,
    state: tauri::State<'_, RecordingState>,
) -> Result<(), String> {
    let mut service = state.service.lock().map_err(|e| e.to_string())?;
    let cfg = config.unwrap_or_default();
    service.start(cfg)
}

#[command]
pub fn stop_recording(state: tauri::State<'_, RecordingState>) -> Result<String, String> {
    let mut service = state.service.lock().map_err(|e| e.to_string())?;
    service.stop()
}

#[command]
pub fn pause_recording(state: tauri::State<'_, RecordingState>) -> Result<(), String> {
    let mut service = state.service.lock().map_err(|e| e.to_string())?;
    service.pause()
}

#[command]
pub fn resume_recording(state: tauri::State<'_, RecordingState>) -> Result<(), String> {
    let mut service = state.service.lock().map_err(|e| e.to_string())?;
    service.resume()
}

#[command]
pub fn get_recording_status(state: tauri::State<'_, RecordingState>) -> String {
    let service = state.service.lock().unwrap();
    match service.get_status() {
        RecordingStatus::Idle => "idle".to_string(),
        RecordingStatus::Starting => "starting".to_string(),
        RecordingStatus::Recording => "recording".to_string(),
        RecordingStatus::Paused => "paused".to_string(),
        RecordingStatus::Stopping => "stopping".to_string(),
        RecordingStatus::Error => "error".to_string(),
    }
}

#[command]
pub fn get_recording_stats(state: tauri::State<'_, RecordingState>) -> RecordingStats {
    let service = state.service.lock().unwrap();
    service.get_stats()
}

#[command]
pub fn get_recording_config(state: tauri::State<'_, RecordingState>) -> RecordingConfig {
    let service = state.service.lock().unwrap();
    service.get_config()
}
